enhancements/orange_okt_proposition.md

```diff
@@ -0,0 +1,326 @@
+
+     Start->Running->Servicing
+                   ->Stopping
+                   -> End
+Going from 1 step to the other is conditionned by some conditions that may be met  **over N Reconciliation cycles**.
+
+I like the idea to have a clear view on the steps I defined previously, so I'll complete my work with the OperatorSDK and the OKT addon.
+
+OKT comes with a statemachine feature that should help in defining these steps and let me focus on the code I need to implement at each step. 
+To allow this, OKT provides:
+
+  -  a sidecar for my application to help me to get my database status and launch actions on it asynchronously.
+  -  an utility to modelize my graph of appication states into my CRD
+  -  a GO type to implement this graph and transition rules that condition how I validate the transition from one step to another
+
+In my CR I set the wished state (i.e. Servicing) I want to reach, while the current application state (i.e. is maintained in the CR status with a new Condition).
+
+Once the application added to the OKT registry (like any other resource), the OKT Reconciler knows that it has to manage this  resource as follow:
+
+  - on Start: Create() it!
+  - on End: Delete() it!
+  - on any other state: Update it!
+
+As any other resource, it put in place an idempotent mecanism and detect changes (and thus will do nothing during a Reconciliation if there's nothing new). Here what will trigger a change:
+
+  - a state change (in App LC Graph) due to a CR modification
+  - a state change from the observation of a change at the application level. This observability should be implemented by an application sidecar container or a usable function in the application container itself. 
+ 
+ A state change (in the App LC  Graph) is handled asynchronously to not impact the Controller with a too long task. On such case (long task) 1 or more requeueing orders are left to wait for the observable change once done. 
+
+It also maintain a Status condition in the CR that reflect the application current state and errors if any.
+
+To sum up:
+
+  - an application lifecycle is managed like an infrastucture resource from OKT's point of view, 
+- a clear view on what is implemented in term of application lifecycle is provided thanks to the App LC Graph described by the CRD
+  - Having all the operators in an organization built upon the same model should help human (or intelligent automates) operators to deal with several kind of K8S operators.
+
+### Implementation Details/Notes/Constraints [optional]
+
+Today, the Story 1 is fully operational and implementable with the current version of OKT. A partial implementation detail can be provided by the MemcachedOperatorSampleWithOKT repo code.
+
+The architecture to render story 2 implementable is not yet fully completed and we are wondering if this approach make sense or not.

```

Comment:
I think the concept of a controller state machine makes a lot of sense, and one generic enough to register an arbitrary application lifecycle for an operator to execute would be powerful. I think such a library warrants an enhancement on its own, with a prototype ideally.

Reply:
Yes thank you, the concept is interresting and I have some ideas to render the LC graph implementation generic at some points. Note that the K8S resources reconciliation steps (CRChecker(), ObjectsGetter(), ..., SuccessManager() ) are already implemented thanks to a StateMachine library ([Qor transition](https://github.com/qor/transition)), it is a bit overrated because initialy I thought about a more complex reconciliation process, but it's done with that right now.
It is clear now, that it is better to separate in 2 distincts processes the K8S resources reconciliation and the Application's LC graph stages. 
Besides that, some ideas comes in mind. We could manage the state changes through a standard events formalizations as the one brought by the [CloudEvents](https://github.com/cloudevents/spec) specs and lib. Providing a standard way to observe and pilot the state change seems interesting.

Reply:
:+1: agree that this state machine idea sounds quite interesting.